File: sast-demo-python/app/crypto_issues.py
"""
crypto_issues.py — Weak cryptography and hardcoded secrets

SAST findings:
  - B105/B106/B107 (Bandit): hardcoded password / secret
  - B303 (Bandit): use of MD5 / SHA1 — weak hash algorithm
  - B304/B305 (Bandit): use of DES / RC4 — weak cipher
  - B311 (Bandit): random.random() used for security purpose
  - SonarQube S2070: SHA-1 and MD5 are not suitable for security-sensitive hashing
  - SonarQube S2245: pseudorandom number generators should not be used in security contexts
"""
import hashlib
import hmac
import random
import string

# pycrypto (pinned to 2.6.1 — CVE-2013-7459 heap overflow)
from Crypto.Cipher import DES, ARC4


# -----------------------------------------------------------------------
# Finding 1: Hardcoded credentials
# -----------------------------------------------------------------------
SECRET_KEY      = "hardcoded-django-secret-key-1234567890abcdef"  # SAST: hardcoded secret
DB_PASSWORD     = "admin123"                                       # SAST: hardcoded password
AWS_SECRET      = "wJalrXUtnFEMI/K7MDENG/bPxRfiCYEXAMPLEKEY"    # SAST: hardcoded AWS key
JWT_SECRET      = "my_jwt_secret"                                  # SAST: hardcoded token


# -----------------------------------------------------------------------
# Finding 2: MD5 used for password hashing
# -----------------------------------------------------------------------
def hash_password_md5(password):
    # SAST: MD5 is cryptographically broken and must not be used for passwords
    return hashlib.md5(password.encode()).hexdigest()


# -----------------------------------------------------------------------
# Finding 3: SHA-1 used for password hashing
# -----------------------------------------------------------------------
def hash_password_sha1(password):
    # SAST: SHA-1 is deprecated for security-sensitive hashing
    return hashlib.sha1(password.encode()).hexdigest()


# -----------------------------------------------------------------------
# Finding 4: DES encryption (weak, 56-bit key)
# -----------------------------------------------------------------------
def encrypt_des(plaintext: bytes) -> bytes:
    key = b"8bytekey"  # SAST: hardcoded key + weak DES cipher
    cipher = DES.new(key, DES.MODE_ECB)  # SAST: ECB mode is insecure
    # Pad to 8-byte boundary
    padded = plaintext + b"\x00" * (8 - len(plaintext) % 8)
    return cipher.encrypt(padded)


# -----------------------------------------------------------------------
# Finding 5: RC4 (ARC4) stream cipher — cryptographically broken
# -----------------------------------------------------------------------
def encrypt_rc4(data: bytes) -> bytes:
    key = b"rc4key"  # SAST: hardcoded key + broken RC4 cipher
    cipher = ARC4.new(key)
    return cipher.encrypt(data)


# -----------------------------------------------------------------------
# Finding 6: Insecure random for security token
# -----------------------------------------------------------------------
def generate_session_token(length=32):
    # SAST: random.choice is not cryptographically secure
    chars = string.ascii_letters + string.digits
    return "".join(random.choice(chars) for _ in range(length))


# -----------------------------------------------------------------------
# Finding 7: Insecure HMAC comparison (timing attack)
# -----------------------------------------------------------------------
def verify_token_insecure(token, expected):
    # SAST: == comparison of secrets is vulnerable to timing attacks
    # Should use hmac.compare_digest()
    return token == expected

File: sast-demo-ts/src/cryptoIssues.ts
/**
 * cryptoIssues.ts — Weak cryptography and hardcoded secrets
 *
 * SAST findings:
 *   - CWE-327: use of broken/weak cryptographic algorithm (MD5, SHA1)
 *   - CWE-330: Math.random() for security-sensitive tokens
 *   - CWE-798: hardcoded credentials / secrets
 *   - CWE-347: JWT accepted without signature verification
 *   - SonarQube S2068, S2245, S4432, S5659
 *
 * Dependency: jsonwebtoken 8.5.1 — CVE-2022-23529 (arbitrary file read)
 */
import * as crypto from "crypto";
import * as jwt from "jsonwebtoken";

// -------------------------------------------------------------------------
// Finding 1-4: Hardcoded secrets
// -------------------------------------------------------------------------
const DB_PASSWORD  = "s3cr3t_db_p@ssword";           // SAST: hardcoded password
const JWT_SECRET   = "my-super-secret-jwt-key";       // SAST: hardcoded JWT secret
const API_KEY      = "AKIAIOSFODNN7EXAMPLE";           // SAST: hardcoded AWS key
const ENCRYPT_KEY  = "0123456789abcdef";               // SAST: hardcoded crypto key

// -------------------------------------------------------------------------
// Finding 5: MD5 used for password hashing
// -------------------------------------------------------------------------
export function hashPasswordMD5(password: string): string {
  // SAST: MD5 is cryptographically broken — must not be used for passwords
  return crypto.createHash("md5").update(password).digest("hex");
}

// -------------------------------------------------------------------------
// Finding 6: SHA-1 used for password hashing
// -------------------------------------------------------------------------
export function hashPasswordSHA1(password: string): string {
  // SAST: SHA-1 is deprecated for security-sensitive hashing
  return crypto.createHash("sha1").update(password).digest("hex");
}

// -------------------------------------------------------------------------
// Finding 7: DES cipher (weak, 56-bit key)
// -------------------------------------------------------------------------
export function encryptDES(plaintext: string): Buffer {
  // SAST: DES is a broken cipher; key size is insufficient
  const cipher = crypto.createCipheriv("des-ecb", Buffer.from("8bytekey"), null);
  return Buffer.concat([cipher.update(plaintext), cipher.final()]);
}

// -------------------------------------------------------------------------
// Finding 8: RC4 (deprecated, broken stream cipher)
// -------------------------------------------------------------------------
export function encryptRC4(plaintext: string): Buffer {
  // SAST: RC4 is cryptographically broken
  const cipher = crypto.createCipheriv("rc4", Buffer.from("rc4keyvalue"), null as any);
  return Buffer.concat([cipher.update(plaintext), cipher.final()]);
}

// -------------------------------------------------------------------------
// Finding 9: Math.random() for session token (not cryptographically secure)
// -------------------------------------------------------------------------
export function generateToken(): string {
  // SAST: Math.random() is predictable — use crypto.randomBytes() instead
  return Math.random().toString(36).substring(2);
}

// -------------------------------------------------------------------------
// Finding 10: JWT signed with hardcoded secret and weak algorithm
// -------------------------------------------------------------------------
export function createToken(userId: string): string {
  // SAST: hardcoded secret + HS256 is weaker than RS256
  return jwt.sign({ userId }, JWT_SECRET, { algorithm: "HS256" });
}

// -------------------------------------------------------------------------
// Finding 11: JWT decoded without verifying signature
// -------------------------------------------------------------------------
export function decodeTokenInsecure(token: string): any {
  // SAST: { complete: true } without verify() — signature never checked
  return jwt.decode(token, { complete: true });
}

// -------------------------------------------------------------------------
// Finding 12: JWT accepted with algorithm:none (CVE-2022-23529 family)
// -------------------------------------------------------------------------
export function verifyTokenInsecure(token: string): any {
  // SAST: algorithms:['none'] disables signature verification entirely
  return jwt.verify(token, "", { algorithms: ["none"] } as any);
}

// -------------------------------------------------------------------------
// Finding 13: Timing-attack-vulnerable secret comparison
// -------------------------------------------------------------------------
export function validateApiKey(provided: string): boolean {
  // SAST: == is susceptible to timing attacks; use crypto.timingSafeEqual()
  return provided === API_KEY;
}
